serve/mlc_serve/model/sampler.py

```diff
@@ -0,0 +1,472 @@
+import torch
+import numpy as np
+import structlog
+from dataclasses import dataclass
+from typing import List, Optional, Tuple
+from ..engine import (
+    SamplingParams,
+    SamplingType,
+    SAMPLING_EPS,
+    LOGPROB_TOP_K_MAX,
+    RawLogprobsInfo,
+)
+
+LOG = structlog.stdlib.get_logger(__name__)
+
+
+def _apply_top_p_top_k(logits, top_ps, top_ks):
+    # TODO(@team): Check the ordering. We currently apply top-p -> top-k.
+    logits_sort, logits_idx = logits.sort(dim=-1, descending=True)
+
+    # Apply top-p.
+    probs_sort = logits_sort.softmax(dim=-1)
+    probs_sum = probs_sort.cumsum(dim=-1)
+    top_p_mask = (probs_sum - probs_sort) > top_ps.unsqueeze(dim=1)
+    logits_sort[top_p_mask] = -float("inf")
+
+    # Apply top-k.
+    # Create a mask for the top-k elements.
+    top_k_mask = torch.arange(logits_idx.shape[-1], device=logits_idx.device)
+    top_k_mask = top_k_mask.expand(logits_idx.shape[0], -1)
+    top_k_mask = top_k_mask >= top_ks.unsqueeze(dim=1)
+    logits_sort[top_k_mask] = -float("inf")
+
+    # Re-sort the probabilities.
+    logits = torch.gather(logits_sort, dim=-1, index=torch.argsort(logits_idx, dim=-1))
+    return logits
+
+
+@dataclass
+class SamplingTensors:
+    mask_random: torch.Tensor
```

Comment:
@masahi we might want to look into https://github.com/patrick-kidger/jaxtyping or another project to also get more checking on the tensors. I think its relatively value-able especially as we have more and more people working on this code.